internal/validators/privileges/role_privileges.go

```diff
@@ -117,11 +109,91 @@ func getUserAndGroupPrincipals(ctx context.Context, username string, driver *vsp
 
 		header.Security, cerr = tokens.Issue(ctx, req)
 		if cerr != nil {
-			return "", nil, cerr
+			return nil, cerr
 		}
 	}
 
 	if err = ssoClient.Login(ssoClient.WithHeader(ctx, header)); err != nil {
+		return nil, err
+	}
+
+	return ssoClient, nil
+}
+
+func isAdminAccount(ctx context.Context, driver *vsphere.VSphereCloudDriver) (bool, error) {
+	ssoClient, err := configureSSOClient(ctx, driver)
+	if err != nil {
+		return false, err
+	}
+	defer ssoClient.Logout(ctx)
+
+	_, err = ssoClient.FindUser(ctx, driver.VCenterUsername)
+	if err != nil {
+		if strings.Contains(err.Error(), "NoPermission") {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func getPrivileges(ctx context.Context, driver *vsphere.VSphereCloudDriver, authManager *object.AuthorizationManager, username string) (map[string]bool, error) {
+	isAdmin, err := IsAdminAccount(ctx, driver)
+	if err != nil {
+		return nil, err
+	}
+
+	groupPrincipals := make([]string, 0)
+	if isAdmin {
+		userPrincipal, groupPrincipals, err := GetUserAndGroupPrincipals(ctx, username, driver)
+		if err != nil {
+			return nil, err
+		}
+
+		privileges, err := vsphere.GetVmwareUserPrivileges(ctx, userPrincipal, groupPrincipals, authManager)
+		if err != nil {
+			return nil, err
+		}
+
+		return privileges, nil
+	}
+
+	userPrincipal, err := driver.GetCurrentVmwareUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !isSameUser(userPrincipal, username) {
+		return nil, errors.New("Not authorized to get privileges for another user from non-admin account")
+	}
+
+	privileges, err := vsphere.GetVmwareUserPrivileges(ctx, userPrincipal, groupPrincipals, authManager)
+	if err != nil {
+		return nil, err
+	}
+
+	return privileges, nil
+}
+
+// checks if a user principle (VSPHERE.LOCAL\username) matches the username (username@vsphere.local)
```

Comment:
nit: the comment should be `VSPHERE.LOCAL\\username`, right?

Reply:
no, the principle is in the form of `VSPHERE.LOCAL\ahmad`. I can update the code to use a literal string so we dont need to escape the `\` to make that more clear

Reply:
Gotcha. Thanks!